flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSubpartitionFileReaderTest.java

```diff
@@ -0,0 +1,418 @@
+        subpartitionFileReader.readBuffers(segments, FreeingBufferRecycler.INSTANCE);
+        assertThat(segments).isEmpty();
+        assertThat(loadedBuffers).hasSize(1);
+
+        bufferIndexOrError = loadedBuffers.poll();
+        assertThat(bufferIndexOrError.getBuffer()).isPresent();
+        assertThat(bufferIndexOrError.getThrowable()).isNotPresent();
+        assertThat(bufferIndexOrError.getIndex()).isEqualTo(4);
+    }
+
+    @Test
+    void testReadBufferNotExceedThreshold() throws Exception {
+        HsSubpartitionFileReader subpartitionFileReader = createSubpartitionFileReader();
+        Deque<BufferIndexOrError> loadedBuffers = subpartitionFileReader.getLoadedBuffers();
+
+        writeDataToFile(targetChannel, 0, MAX_BUFFERS_READ_AHEAD + 1);
+
+        subpartitionFileReader.prepareForScheduling();
+        // allocate maxBuffersReadAhead + 1 read buffers for reading
+        Queue<MemorySegment> memorySegments = createsMemorySegments(MAX_BUFFERS_READ_AHEAD + 1);
+        // trigger reading
+        subpartitionFileReader.readBuffers(memorySegments, FreeingBufferRecycler.INSTANCE);
+
+        // preload by fileReader will not exceed threshold
+        assertThat(loadedBuffers).hasSize(MAX_BUFFERS_READ_AHEAD);
+        assertThat(memorySegments).hasSize(1);
+    }
+
+    @Test
+    void testReadBufferNotifyDataAvailable() throws Exception {
+        HsSubpartitionFileReader subpartitionFileReader = createSubpartitionFileReader();
+        BlockingDeque<BufferIndexOrError> loadedBuffers =
+                (BlockingDeque<BufferIndexOrError>) subpartitionFileReader.getLoadedBuffers();
+
+        writeDataToFile(targetChannel, 0, 1);
+        subpartitionFileReader.prepareForScheduling();
+        Queue<MemorySegment> memorySegments = createsMemorySegments(1);
+        subpartitionFileReader.readBuffers(memorySegments, FreeingBufferRecycler.INSTANCE);
+        assertThat(subpartitionOperation.getNotifyNum()).isEqualTo(1);
+        assertThat(loadedBuffers).hasSize(1);
+
+        // trigger next round reading.
+        final int numBuffers = MAX_BUFFERS_READ_AHEAD;
+        subpartitionOperation.advanceConsumptionProgress();
+        writeDataToFile(targetChannel, 1, numBuffers);
+        subpartitionFileReader.prepareForScheduling();
+
+        OneShotLatch consumerThreadLatch = new OneShotLatch();
+        AtomicReference<Throwable> cause = new AtomicReference<>(null);
+        new Thread(
+                        () -> {
+                            consumerThreadLatch.trigger();
+                            for (int i = 0; i < numBuffers; i++) {
+                                try {
+                                    loadedBuffers.take();
+                                } catch (InterruptedException e) {
+                                    cause.set(e);
+                                }
+                            }
+                        })
+                .start();
+        consumerThreadLatch.await();
```

Comment:
Could you explain a bit more about this logic?

Reply:
This part is refactored to test that the consumer thread can consume all buffers in time.